azurelinuxagent/ga/utils/exthandler_utils.py

```diff
@@ -0,0 +1,343 @@
+DEFAULT_MEM_LIMIT_MAX_MB_FOR_AGENT = 512.0  # mb, applies to agent only
+DEFAULT_MEM_LIMIT_PCT_FOR_EXTN = 15.0  # percent, applies to extensions
+
+CGROUPS_ENFORCE_DEFAULT_LIMITS = False  # Currently, we don't want to enforce defaults as well.
+
+
+class HandlerConfiguration(object):
+    @staticmethod
+    def send_handler_configuration_event(name=AGENT_NAME, message="", operation=WALAEventOperation.Unknown,
+                                         is_success=True, log_event=True):
+        if message:
+            add_event(name=name, message=message, op=operation, is_success=is_success, log_event=log_event)
+
+    def __init__(self, handler_config_json, name=None):
+        self._data = handler_config_json
+        self._name = name
+        self.resource_config = None
+        self.resource_limits = None
+
+        try:
+            if handler_config_json is None:
+                raise ExtensionHandlerConfigurationError('Malformed handler configuration file. '
+                                                  'No data present')
+            if "handlerConfiguration" not in handler_config_json:
+                raise ExtensionHandlerConfigurationError('Malformed handler configuration file. '
+                                                  '"HandlerConfiguration" key not present')
+            if "linux" not in handler_config_json['handlerConfiguration']:
+                raise ExtensionHandlerConfigurationError('No linux configurations present in '
+                                                  'HandlerConfiguration')
+
+            try:
+                self.resource_config = ExtensionResourcesConfiguration(self.get_linux_configurations())
+            except ExtensionHandlerConfigurationError as e:
+                raise e
+
+            if self.resource_config or CGROUPS_ENFORCE_DEFAULT_LIMITS:
+                self.resource_limits = CGroupsLimits(self._name, self.resource_config)
+
+        except ExtensionHandlerConfigurationError as e:
+            logger.warn(ustr(e))
+            self.send_handler_configuration_event(message=ustr(e), is_success=False, log_event=False,
+                                                                  operation=WALAEventOperation.HandlerConfiguration)
+
+    def get_name(self):
+        return self._data.get("name", None)
+
+    def get_version(self):
+        return self._data.get("version", None)
+
+    def get_linux_configurations(self):
+        return self._data['handlerConfiguration']['linux']
+
+    def get_resource_configurations(self):
+        return self.resource_config
+
+    def get_resource_limits(self):
+        return self.resource_limits
+
+
+class ExtensionResourcesConfiguration(object):
+    def __init__(self, resource_configuration):
+        self._data = resource_configuration["resources"]
+        self.cpu_limits = None
+        self.memory_limits = None
+
+        try:
+            if "cpu" not in self._data and "memory" not in self._data:
```

Comment:
i think the validation needs to be stricter... this only checks for the "cpu" or "memory" substring to be anywhere in the json text

Reply:
never mind - it's a dict